vllm/model_executor/layers/sampler.py

```diff
@@ -769,33 +769,37 @@ def _get_logprobs(
     selected_logprobs = logprobs[[
         query_indices_gpu,
         next_token_ids_gpu,
-    ]]
+    ]].to('cpu')
     ranks = _get_ranks(
         logprobs[query_indices_gpu],
         next_token_ids_gpu,
-    )
-    assert selected_logprobs.shape[0] == ranks.shape[0]
+    ).to('cpu')
 
     # Logprobs of topk tokens for a batch of sequence groups.
     # (num_query_tokens_across_batch).
     if largest_num_logprobs > 0:
         top_logprobs, top_token_ids = torch.topk(logprobs,
                                                  largest_num_logprobs,
                                                  dim=-1)
-        top_logprobs = top_logprobs.cpu()
-        top_token_ids = top_token_ids.cpu()
     else:
         top_logprobs, top_token_ids = None, None
 
-    selected_logprobs = selected_logprobs.cpu()
-    ranks = ranks.cpu()
+    # Overlap cpu transfer with gpu kernels.
+    selected_logprobs = selected_logprobs.to('cpu')
+    ranks = ranks.to('cpu')
+    if top_logprobs is not None and top_token_ids is not None:
+        top_logprobs = top_logprobs.to('cpu')
+        top_token_ids = top_token_ids.to('cpu')
 
     # Find prompt/sample logprobs.
     prompt_logprobs_per_seq_group: List[Optional[PromptLogprobs]] = []
     sample_logprobs_per_seq_group: List[SampleLogprobs] = []
     top_logprob_idx = 0
     selected_logprobs_idx = 0
 
+    # Make sure non-blocking .to("cpu", non_blocking=True) is finished
+    assert selected_logprobs.shape[0] == ranks.shape[0]
```

Comment:
@rkooo567 Do we still need this assert since non-blocking transfer code is removed?